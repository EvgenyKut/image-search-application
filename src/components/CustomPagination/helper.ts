export const paginationFormatter = (
  page: number,
  totalPages: number,
  type: 'increase' | 'decrease',
  value: number = 1,
) => {
  const positiveValue = page + value > totalPages ? totalPages : page + value;
  const negativeValue = page - value < 0 ? 0 : page - value;
  return type === 'increase' ? positiveValue : negativeValue;
};
