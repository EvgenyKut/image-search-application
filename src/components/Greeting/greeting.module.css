.wrapper {
  display: flex;
  font-size: 20px;
  align-items: center;
}

.avatar {
  margin-right: 0.5rem;
}

.lightTheme {
  display: none;
}

.darkTheme {
  display: none;
}

body[data-theme='light'] .lightTheme {
  display: block;
}

body[data-theme='dark'] .darkTheme {
  display: block;
}
