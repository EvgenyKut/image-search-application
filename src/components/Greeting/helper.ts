export const greetingFormatting = (name: string, time: number) => {
  let timePeriod = '';

  if (time >= 22 || time < 6) {
    timePeriod = 'night';
  } else if (time >= 18) {
    timePeriod = 'evening';
  } else if (time >= 12) {
    timePeriod = 'afternoon';
  } else if (time >= 6) {
    timePeriod = 'morning';
  }
  return name ? `Good ${timePeriod}, ${name}!` : `Good ${timePeriod}!`;
};
