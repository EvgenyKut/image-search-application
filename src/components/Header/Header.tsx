import React from 'react';
import { useNavigate } from 'react-router-dom';
import Button from '@mui/material/Button';
import { Greeting } from '../../components';
import style from './header.module.css';
import { useAppDispatch, useAppSelector } from '../../hooks';
import { resetName } from '../../store/reducers/AuthSlice';
import { Dictionary } from '../../constants/dictionary';

type HeaderProps = { children: React.ReactNode };
const Header: React.FC<HeaderProps> = ({ children }) => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const { name } = useAppSelector((state) => state.authReducer);

  const logoutHandler = () => {
    dispatch(resetName());
    navigate('/');
    localStorage.removeItem('search-app-name');
  };

  return (
    <div className={style.wrapper}>
      <div className={style.container}>
        <Greeting name={name} />
        {children}
      </div>
      {name && (
        <Button
          onClick={logoutHandler}
          variant="contained"
          color="primary"
          type="button"
        >
          {Dictionary.LOGOUT}
        </Button>
      )}
    </div>
  );
};

export default Header;
