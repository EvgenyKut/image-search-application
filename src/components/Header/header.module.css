.wrapper {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1rem;
  background-color: lightblue;
}

.container {
  display: flex;
}
