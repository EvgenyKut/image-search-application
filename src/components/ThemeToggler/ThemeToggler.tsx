import React from 'react';
import { useAppSelector } from '../../hooks';
import useKeyListener from '../../hooks/useKeyListener';
import useLocalStorageState from '../../hooks/useLocalStorageState';
import { ReactComponent as Moon } from '../../assets/Moon.svg';
import { ReactComponent as Sun } from '../../assets/Sun.svg';
import styles from './themeToggler.module.css';

const ThemeToggler = () => {
  const { focus } = useAppSelector((state) => state.searchFocusReducer);
  const [theme, setTheme] = useLocalStorageState('search-app-theme', 'light');
  const nextTheme = theme === 'light' ? 'dark' : 'light';

  const setNextTheme = () => setTheme(nextTheme);
  useKeyListener('KeyF', setNextTheme, focus);

  React.useEffect(() => {
    document.body.dataset.theme = theme;
  }, [theme]);

  return (
    <div
      onClick={setNextTheme}
      className={styles.wrapper}
      data-testid="theme-toggler"
    >
      <div data-testid="dark" className={styles.darkTheme}>
        <Moon />
      </div>
      <div data-testid="light" className={styles.lightTheme}>
        <Sun />
      </div>
    </div>
  );
};

export default ThemeToggler;
