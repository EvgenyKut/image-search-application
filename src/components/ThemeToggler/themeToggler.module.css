.wrapper {
  display: flex;
  align-items: center;
  cursor: pointer;
  margin: 0 1rem;
}

.lightTheme {
  display: none;
}

.darkTheme {
  display: none;
}

body[data-theme='light'] .lightTheme {
  display: block;
}

body[data-theme='dark'] .darkTheme {
  display: block;
}
