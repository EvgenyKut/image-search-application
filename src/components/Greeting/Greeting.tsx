import React from 'react';
import styles from './greeting.module.css';
import { greetingFormatting } from './helper';
import Avatar from '@mui/material/Avatar';
import { useAppSelector } from '../../hooks';
import { capitalizeFirstLetter } from '../../helpers/helper';

type GreetingProps = {
  name: string;
};
const Greeting: React.FC<GreetingProps> = ({ name }) => {
  const { isAuth } = useAppSelector((state) => state.authReducer);
  return (
    <div className={styles.wrapper} data-testid="greeting">
      {isAuth && (
        <Avatar className={styles.avatar}>
          {capitalizeFirstLetter(name.slice(0, 1))}
        </Avatar>
      )}
      {greetingFormatting(name, new Date().getHours())}
    </div>
  );
};

export default Greeting;
