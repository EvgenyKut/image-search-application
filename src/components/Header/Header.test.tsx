import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import Header from './index';
import * as reactRedux from 'react-redux';
import { useNavigate } from 'react-router-dom';
import ThemeToggler from '../ThemeToggler';
import { greetingFormatting } from '../Greeting/helper';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(() => jest.fn),
}));

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('Header', () => {
  const navigate = jest.fn();
  const useSelectorMock = reactRedux.useSelector;
  const useNavigateMock = useNavigate;

  const mockStore = {
    authReducer: {
      name: '',
      isAuth: false,
    },
    searchFocusReducer: {
      focus: false,
    },
  };

  beforeEach(() => {
    (useSelectorMock as jest.Mock).mockImplementation((selector) =>
      selector(mockStore),
    );
    (useNavigateMock as jest.Mock).mockImplementation(() => navigate);
  });

  afterEach(() => {
    (useSelectorMock as jest.Mock).mockClear();
    (useNavigateMock as jest.Mock).mockClear();
  });

  it('snapshot', () => {
    const tree = renderer
      .create(
        <Header>
          <div>test</div>
        </Header>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Header (not auth)', () => {
    render(
      <Header>
        <ThemeToggler />
      </Header>,
    );
    const toggler = screen.getByTestId('theme-toggler');
    expect(toggler).toBeInTheDocument();
    expect(toggler).toBeVisible();

    const lightMode = screen.getByTestId('light');
    expect(lightMode).toBeInTheDocument();
    expect(lightMode).toHaveStyle({ display: 'block' });

    const greeting = screen.getByTestId('greeting');
    expect(greeting).toBeInTheDocument();
    const value = greetingFormatting('', new Date().getHours());
    expect(greeting.textContent).toBe(value);
  });
});

describe('Header, auth', () => {
  const navigate = jest.fn();
  const useDispatchMock = reactRedux.useDispatch;
  const useSelectorMock = reactRedux.useSelector;
  const useNavigateMock = useNavigate;

  const userName = 'Evgeny';

  const mockStore = {
    authReducer: {
      name: userName,
      isAuth: true,
    },
    searchFocusReducer: {
      focus: false,
    },
  };

  beforeEach(() => {
    (useDispatchMock as jest.Mock).mockImplementation(() => () => {}); // eslint-disable-line
    (useSelectorMock as jest.Mock).mockImplementation((selector) =>
      selector(mockStore),
    );
    (useNavigateMock as jest.Mock).mockImplementation(() => navigate);
  });

  afterEach(() => {
    (useDispatchMock as jest.Mock).mockClear();
    (useSelectorMock as jest.Mock).mockClear();
    (useNavigateMock as jest.Mock).mockClear();
  });

  it('Header auth behavior', () => {
    render(
      <Header>
        <ThemeToggler />
      </Header>,
    );
    const toggler = screen.getByTestId('theme-toggler');
    expect(toggler).toBeInTheDocument();
    expect(toggler).toBeVisible();

    const lightMode = screen.getByTestId('light');
    expect(lightMode).toBeInTheDocument();
    expect(lightMode).toHaveStyle({ display: 'block' });

    const greeting = screen.getByTestId('greeting');
    expect(greeting).toBeInTheDocument();
    const value = greetingFormatting(userName, new Date().getHours());
    expect(greeting.textContent).toBe(value);

    const logoutBtn = screen.getByRole('button');
    expect(logoutBtn.textContent).toBe('Logout');
    expect(logoutBtn).toBeInTheDocument();
    expect(logoutBtn).toBeVisible();

    fireEvent.click(logoutBtn);
    expect(navigate).toHaveBeenLastCalledWith('/');
  });
});
