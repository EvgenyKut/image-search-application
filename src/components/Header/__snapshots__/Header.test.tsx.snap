// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot 1`] = `
<div
  className="wrapper"
>
  <div
    className="container"
  >
    <div
      className="wrapper"
      data-testid="greeting"
    >
      Good morning!
    </div>
    <div>
      test
    </div>
  </div>
</div>
`;
